src/main/java/xyz/kyngs/herbot/handlers/command/AbstractCommand.java

```diff
@@ -1,17 +1,119 @@
 package xyz.kyngs.herbot.handlers.command;
 
+import net.dv8tion.jda.api.EmbedBuilder;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.entities.TextChannel;
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
 import xyz.kyngs.herbot.HerBot;
+import xyz.kyngs.herbot.handlers.user.UserProfile;
+import xyz.kyngs.herbot.util.Argument;
+import xyz.kyngs.herbot.util.ClassUtil;
+
+import java.awt.*;
+import java.util.ArrayList;
+import java.util.List;
+
+import static xyz.kyngs.herbot.util.ExceptionUtil.completedExceptionally;
 
 public abstract class AbstractCommand implements CommandExecutor {
 
     protected final HerBot herBot;
     protected final String description;
+    @SuppressWarnings("rawtypes")
+    private final List<Argument> args;
+    private final String permission;
 
-    public AbstractCommand(HerBot herBot, String description) {
+    public AbstractCommand(HerBot herBot, String description, String permission) {
         this.herBot = herBot;
         this.description = description;
+        this.permission = permission;
+        args = new ArrayList<>();
+    }
+
+    public <T> void addArg(Class<T> type, String... names) {
+        args.add(new Argument<T>(type, names));
     }
 
+    @Override
+    public void onCommand(User author, Guild guild, TextChannel channel, Message message, String[] args, UserProfile profile, GuildMessageReceivedEvent event) {
```

Comment:
This function is too long. I would split it into smaller functions, which might be useful later if we decide to go for the unit tests.

Reply:
oncommand is basically a listener for when command happens, so it needs to be one function

Reply:
But it shouldn't change a thing if you fopr example take the lines 80-104 into a new function.
something like
`private isEverythingOk(String[] args) {
     bool allOk = true;
         for (int i = 0; i < this.args.size(); i++) {
             allOk = allOk and lookingGood; // Not sure how you write logical and in the Java.
         }
}`
And then you just call this function instead of cycle and it should work the same way.

Reply:
I think I can do that, however, I also think it is redundant.